compiler/qsc_formatter/src/formatter.rs

```diff
@@ -0,0 +1,254 @@
+    pub new_text: String,
+}
+
+impl Edit {
+    fn new(lo: u32, hi: u32, new_text: &str) -> Self {
+        Self {
+            span: Span { lo, hi },
+            new_text: new_text.to_string(),
+        }
+    }
+}
+
+fn make_indent_string(level: usize) -> String {
+    "    ".repeat(level)
+}
+
+/// Applies formatting rules to the given code str, generating edits where
+/// the source code needs to be changed to comply with the format rules.
+pub fn format(code: &str) -> Vec<Edit> {
+    let tokens = concrete::ConcreteTokenIterator::new(code);
+    let mut edits = vec![];
+
+    let mut indent_level: usize = 0;
+
+    // The sliding window used is over three adjacent tokens
+    #[allow(unused_assignments)]
+    let mut one = None;
+    let mut two = None;
+    let mut three = None;
+
+    for token in tokens {
+        // Advance the token window
+        one = two;
+        two = three;
+        three = Some(token);
+
+        let mut edits_for_triple = match (&one, &two, &three) {
+            (Some(one), Some(two), Some(three)) => {
+                // if the token is a {, increase the indent level
+                if let ConcreteTokenKind::Syntax(TokenKind::Open(Delim::Brace)) = one.kind {
+                    indent_level += 1;
+                }
+
+                // if the token is a }, decrease the indent level
+                if let ConcreteTokenKind::Syntax(TokenKind::Close(Delim::Brace)) = one.kind {
+                    indent_level = indent_level.saturating_sub(1);
+                }
+
+                if let ConcreteTokenKind::WhiteSpace = one.kind {
+                    // first token is whitespace, continue scanning
+                    continue;
+                } else if let ConcreteTokenKind::WhiteSpace = two.kind {
+                    // whitespace in the middle
+                    apply_rules(
+                        one,
+                        get_token_contents(code, two),
+                        three,
+                        code,
+                        indent_level,
+                    )
+                } else {
+                    // one, two are adjacent tokens with no whitespace in the middle
+                    apply_rules(one, "", two, code, indent_level)
+                }
+            }
+            (None, None, Some(three)) => {
+                // Remove any whitespace at the start of a file
+                if three.span.lo != 0 {
+                    vec![Edit::new(0, three.span.lo, "")]
+                } else {
+                    vec![]
+                }
+            }
+            _ => {
+                // not enough tokens to apply a rule
+                continue;
+            }
+        };
+
+        edits.append(&mut edits_for_triple);
+    }
+
+    edits
+}
+
+fn fix_whitespace(whitespace: &str, indent_level: usize) -> String {
+    //
+    // when you see newline, insert the indent string
+    // and trim until the next newline or the end of the string
+    //
```

Comment:
This docstring should be before the function

```rust
/// When seeing a new line, insert the indent string and
/// trim until the next newline or the end of the string.
fn fix_whitespace(whitespace: &str, indent_level: usize) -> String { ... }
```

Reply:
function was removed